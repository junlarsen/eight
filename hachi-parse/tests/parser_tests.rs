use hachi_parse::Lexer;
use hachi_parse::Parser;
use insta::assert_ron_snapshot;

#[test]
fn test_snapshot_corpus() {
    insta::glob!("data/*.test", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let mut parser = Parser::new(Lexer::new(&input));
        let translation_unit = parser
            .parse()
            .unwrap_or_else(|_| panic!("failed to parse corpus file {} into ast", path.display()));
        assert_ron_snapshot!(*translation_unit);
    })
}
