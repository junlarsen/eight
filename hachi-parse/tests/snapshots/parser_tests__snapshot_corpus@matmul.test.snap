---
source: hachi-parse/tests/parser_tests.rs
expression: "*translation_unit"
input_file: hachi-parse/tests/data/matmul.test
snapshot_kind: text
---
TranslationUnit(
  items: [
    Type(TypeItem(
      span: Span(
        low: 0,
        high: 50,
      ),
      name: Identifier(
        name: "Matrix",
        span: Span(
          low: 5,
          high: 11,
        ),
      ),
      members: [
        TypeMemberItem(
          span: Span(
            low: 18,
            high: 25,
          ),
          name: Identifier(
            name: "r",
            span: Span(
              low: 18,
              high: 19,
            ),
          ),
          type: Integer32(Integer32Type(
            span: Span(
              low: 21,
              high: 24,
            ),
          )),
        ),
        TypeMemberItem(
          span: Span(
            low: 28,
            high: 35,
          ),
          name: Identifier(
            name: "c",
            span: Span(
              low: 28,
              high: 29,
            ),
          ),
          type: Integer32(Integer32Type(
            span: Span(
              low: 31,
              high: 34,
            ),
          )),
        ),
        TypeMemberItem(
          span: Span(
            low: 38,
            high: 48,
          ),
          name: Identifier(
            name: "buf",
            span: Span(
              low: 38,
              high: 41,
            ),
          ),
          type: Pointer(PointerType(
            span: Span(
              low: 43,
              high: 47,
            ),
            inner: Integer32(Integer32Type(
              span: Span(
                low: 44,
                high: 47,
              ),
            )),
          )),
        ),
      ],
    )),
    Function(FunctionItem(
      span: Span(
        low: 52,
        high: 461,
      ),
      name: Identifier(
        name: "matrix_matrix_multiply",
        span: Span(
          low: 55,
          high: 77,
        ),
      ),
      parameters: [
        FunctionParameterItem(
          span: Span(
            low: 78,
            high: 87,
          ),
          name: Identifier(
            name: "a",
            span: Span(
              low: 78,
              high: 79,
            ),
          ),
          type: Named(NamedType(
            span: Span(
              low: 81,
              high: 87,
            ),
            name: Identifier(
              name: "Matrix",
              span: Span(
                low: 81,
                high: 87,
              ),
            ),
          )),
        ),
        FunctionParameterItem(
          span: Span(
            low: 89,
            high: 98,
          ),
          name: Identifier(
            name: "b",
            span: Span(
              low: 89,
              high: 90,
            ),
          ),
          type: Named(NamedType(
            span: Span(
              low: 92,
              high: 98,
            ),
            name: Identifier(
              name: "Matrix",
              span: Span(
                low: 92,
                high: 98,
              ),
            ),
          )),
        ),
      ],
      return_type: Some(Named(NamedType(
        span: Span(
          low: 102,
          high: 108,
        ),
        name: Identifier(
          name: "Matrix",
          span: Span(
            low: 102,
            high: 108,
          ),
        ),
      ))),
      body: [],
    )),
  ],
)
